java/arcs/android/sdk/host/ArcHostHelper.kt

```diff
@@ -0,0 +1,394 @@
+                it?.let { list ->
+                    list as List<*>
+                    list.map {
+                        (it as ParcelableParticleIdentifier).actual
+                    }
+                }
+            } ?: emptyList()
+
+        override suspend fun startArc(partition: PlanPartition) {
+            sendIntentToArcHostServiceForResult(
+                partition.createStartArcHostIntent(
+                    arcHostComponentName
+                )
+            )
+        }
+
+        override suspend fun stopArc(partition: PlanPartition) {
+            sendIntentToArcHostServiceForResult(
+                partition.createStopArcHostIntent(
+                    arcHostComponentName
+                )
+            )
+        }
+
+        /**
+         * Asynchronously send an [ArcHost] command via [Intent] without waiting for return result.
+         */
+        private fun sendIntentToArcHostService(intent: Intent) {
+            sender(intent)
+        }
+
+        @UseExperimental(ExperimentalCoroutinesApi::class)
+        class ResultReceiverContinuation<T>(
+            val continuation: CancellableContinuation<T?>,
+            val block: (Any?) -> T?
+        ) : ResultReceiver(Handler()) {
+            override fun onReceiveResult(resultCode: Int, resultData: Bundle?) =
+                continuation.resume(
+                    block(
+                        resultData?.get(
+                            OPERATION_RESULT
+                        )
+                    )
+                ) {}
+        }
+
+        /**
+         * Sends an asynchronous [ArcHost] command via [Intent] to a [Service] and waits for a
+         * result using a suspendable coroutine.
+         * @property intent the [ArcHost] command, usually from [ArcHostHelper]
+         * @property transformer a lambda to map return values from a [Bundle] into other types.
+         */
+        private suspend fun <T> sendIntentToArcHostServiceForResult(
+            intent: Intent,
+            transformer: (Any?) -> T?
+        ): T? = suspendCancellableCoroutine { cancelableContinuation ->
+            onResult(
+                intent, ResultReceiverContinuation(cancelableContinuation, transformer)
+            )
+            sendIntentToArcHostService(intent)
+        }
+
+        /**
+         * Sends an asynchronous [ArcHost] command via [Intent] and waits for it to complete
+         * with no return value.
+         */
+        private suspend fun sendIntentToArcHostServiceForResult(
+            intent: Intent
+        ): Unit? = sendIntentToArcHostServiceForResult(intent) {}
+
+        override fun hashCode(): Int = arcHostComponentName.hashCode()
+
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (javaClass != other?.javaClass) return false
+            other as ArcHost
+            if (hostId().equals(other.hostId())) return false
+            return true
+        }
```

Comment:
Doesn't look like `hashCode()` and `equals(other)` use the same attributes when calculating their return values. This could be fraught.